src/SpinWaveTheory/SWTCalculations.jl

```diff
@@ -59,60 +59,18 @@ function swt_hamiltonian!(swt::SpinWaveTheory, k̃ :: Vector{Float64}, Hmat::Mat
     # pairexchange interactions
     for matom = 1:Nm
         ints = sys.interactions_union[matom]
-        # Heisenberg exchange
-        for (; isculled, bond, J) in ints.heisen
-            isculled && break
-            sub_i, sub_j, ΔRδ = bond.i, bond.j, bond.n
-
-            tTi_μ = s̃_mat[:, :, :, sub_i]
-            tTj_ν = s̃_mat[:, :, :, sub_j]
-            phase  = exp(2im * π * dot(k̃, ΔRδ))
-            cphase = conj(phase)
-            sub_i_M1, sub_j_M1 = sub_i - 1, sub_j - 1
-
-            for m = 2:N
-                mM1 = m - 1
-                T_μ_11 = conj(tTi_μ[1, 1, :])
-                T_μ_m1 = conj(tTi_μ[m, 1, :])
-                T_μ_1m = conj(tTi_μ[1, m, :])
-                T_ν_11 = tTj_ν[1, 1, :]
-
-                for n = 2:N
-                    nM1 = n - 1
-                    δmn = δ(m, n)
-                    T_μ_mn, T_ν_mn = conj(tTi_μ[m, n, :]), tTj_ν[m, n, :]
-                    T_ν_n1 = tTj_ν[n, 1, :]
-                    T_ν_1n = tTj_ν[1, n, :]
-
-                    c1 = J * dot(T_μ_mn - δmn * T_μ_11, T_ν_11)
-                    c2 = J * dot(T_μ_11, T_ν_mn - δmn * T_ν_11)
-                    c3 = J * dot(T_μ_m1, T_ν_1n)
-                    c4 = J * dot(T_μ_1m, T_ν_n1)
-                    c5 = J * dot(T_μ_m1, T_ν_n1)
-                    c6 = J * dot(T_μ_1m, T_ν_1n)
 
-                    Hmat11[sub_i_M1*Nf+mM1, sub_i_M1*Nf+nM1] += 0.5 * c1
-                    Hmat11[sub_j_M1*Nf+mM1, sub_j_M1*Nf+nM1] += 0.5 * c2
-                    Hmat22[sub_i_M1*Nf+nM1, sub_i_M1*Nf+mM1] += 0.5 * c1
-                    Hmat22[sub_j_M1*Nf+nM1, sub_j_M1*Nf+mM1] += 0.5 * c2
-
-                    Hmat11[sub_i_M1*Nf+mM1, sub_j_M1*Nf+nM1] += 0.5 * c3 * phase
-                    Hmat22[sub_j_M1*Nf+nM1, sub_i_M1*Nf+mM1] += 0.5 * c3 * cphase
-                    
-                    Hmat22[sub_i_M1*Nf+mM1, sub_j_M1*Nf+nM1] += 0.5 * c4 * phase
-                    Hmat11[sub_j_M1*Nf+nM1, sub_i_M1*Nf+mM1] += 0.5 * c4 * cphase
+        for coupling in ints.pair
+            (; isculled, bond) = coupling
+            isculled && break
 
-                    Hmat12[sub_i_M1*Nf+mM1, sub_j_M1*Nf+nM1] += 0.5 * c5 * phase
-                    Hmat12[sub_j_M1*Nf+nM1, sub_i_M1*Nf+mM1] += 0.5 * c5 * cphase
-                    Hmat21[sub_i_M1*Nf+mM1, sub_j_M1*Nf+nM1] += 0.5 * c6 * phase
-                    Hmat21[sub_j_M1*Nf+nM1, sub_i_M1*Nf+mM1] += 0.5 * c6 * cphase
-                end
+            ### Bilinear exchange
+            if coupling.bilin isa Number
+                J = Mat3(coupling.bilin*I)
+            else
+                J = coupling.bilin
```

Comment:
It seems like a problem to me that these lines are just removed now? Is there supposed to be further changes re-introducing the behavior?

Reply:
The behavior should be the same (if J is stored as a scalar, it has the meaning of a scalar times the 3x3 identity matrix). The priority here is code de-duplication. Performance optimization will be a separate pass.